splight_lib/models/server.py

```diff
@@ -0,0 +1,208 @@
+from datetime import datetime
+from enum import auto
+from typing import Any, NamedTuple
+
+from pydantic import AnyUrl, BaseModel, Field, computed_field
+from strenum import LowercaseStrEnum, PascalCaseStrEnum
+
+from splight_lib.models.database_base import SplightDatabaseBaseModel
+from splight_lib.models.exceptions import (
+    InvalidArgument,
+    InvalidServerConfigType,
+)
+from splight_lib.models.file import File
+from splight_lib.models.hub_server import HubServer
+from splight_lib.models.secret import Secret
+
+warnings.filterwarnings("ignore", category=RuntimeWarning)
+
+NATIVE_TYPES = {
+    "int": int,
+    "bool": bool,
+    "str": str,
+    "float": float,
+    "datetime": datetime,
+    "url": AnyUrl,
+}
+
+DATABASE_TYPES = {
+    "File": File,
+}
+
+DB_MODEL_TYPE_MAPPING = {
+    **NATIVE_TYPES,
+    **DATABASE_TYPES,
+}
+
+
+def get_field_value(field: dict):
+    field_multiple = field.get("multiple", False)
+    field_type = field.get("type")
+    field_value = field.get("value")
+
+    if not field_value:
+        return [] if field_multiple else None
+
+    if field_type in NATIVE_TYPES:
+        value = (
+            field_value
+            if not isinstance(field_value, str)
+            else parse_variable_string(field_value)
+        )
+    elif field_type in DATABASE_TYPES:
+        model_class = DATABASE_TYPES[field_type]
+        value = (
+            model_class.retrieve(field_value)
+            if not field_multiple
+            else [model_class.retrieve(item) for item in field_value]
+        )
+    return value
+
+
+def get_model_class(config: BaseModel, name: str) -> NamedTuple:
+    config_class = namedtuple(name, [x.name for x in config])
+    return config_class
+
+
+def load_server_config(
+    config: list[dict], config_class: NamedTuple
+) -> NamedTuple:
+    config_dict = {}
+    for item in config:
+        if item["type"] not in DB_MODEL_TYPE_MAPPING:
+            raise InvalidServerConfigType(item["name"], item["type"])
+        config_dict.update({item["name"]: get_field_value(item)})
+    config = config_class(**config_dict)
+    return config
+
+
+def load_server_ports(
+    ports: list[dict], model_class: NamedTuple
+) -> namedtuple:
+    ports_dict = {}
+    for port in ports:
+        if isinstance(port["internal_port"], int) and not (
+            0 <= port["internal_port"] <= 65535
+        ):
+            raise ValueError(
+                f"Internal port {port['internal_port']} is out of valid range (0-65535)"
+            )
+        if isinstance(port["exposed_port"], int) and not (
+            0 <= port["exposed_port"] <= 65535
+        ):
+            raise ValueError(
+                f"External port {port['exposed_port']} is out of valid range (0-65535)"
+            )
+        if port["protocol"] not in ["tcp", "udp"]:
+            raise ValueError(
+                f"Protocol {port['protocol']} is not valid. Must be 'tcp' or 'udp'"
+            )
```

Comment:
Use an Enum for the `protocol` field in the `Port` model so this validation is done by pydantic